### PR TITLE
#37 update validators feedback and project

### PR DESCRIPTION
**Pull request by alesyalis**

### CODE REVIEW

File: Recrutify/Recrutify.DataAccess/Repositories/Abstract/ICandidateRepository.cs [hunk 1]
```diff
@@ -7,5 +7,9 @@ namespace Recrutify.DataAccess.Repositories.Abstract
     public interface ICandidateRepository : IBaseRepository<Candidate>
     {
         Task UpsertFeedbackAsync(Guid id, Guid projectId, Feedback feedback);
+
+        Task<ProjectResult> GetCandidateWithProject(Guid id, Guid projectId);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/Abstract/ICandidateRepository.cs:11]: rename GetCandidateWithProjectResult

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs [hunk 2]
```diff
@@ -17,6 +18,34 @@ public CandidateRepository(IOptions<MongoSettings> options)
...
+                       Status = p.Status,
+                       Feedbacks = p.Feedbacks,
+                       ProjectId = p.ProjectId,
+                       Reason = p.Reason,
+                   }).FirstOrDefaultAsync();
+        }
+
+        public Task<ProjectResult> GetCandidateWithProjectFeedbackAsync(Guid id, Guid projectId, Guid feedbackUserId, FeedbackType feedbackType)
+        {
+            return GetCollection().Aggregate().Match(c => c.Id == id)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:37]: .Match from new line, and fix in both methods other places with dot too

File: Recrutify/Recrutify.Services/Validators/UpdateProjectValidator.cs [hunk 3]
```diff
@@ -8,7 +9,7 @@ public class UpdateProjectValidator : BaseProjectValidator<ProjectDTO>
         public UpdateProjectValidator()
         {
             RuleFor(p => p.Id)
-                .NotEmpty();
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/UpdateProjectValidator.cs:11]: return NotEmpty, it's the same as .NotEqual(Guid.Empty)

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            RuleFor(f => f.Feedbacks.FirstOrDefault())
+                .NotNull()
+                .Must(BeAValidData)
+                .WithMessage("Time for editing is over");
+            RuleFor(f => f.Status).IsInEnum()
+                .Must(s => !new[] { Status.Accepted, Status.Declined, Status.WaitingList, }.Contains(s))
+                .WithMessage("Status is not available for update");
+        }
+
+        protected bool BeAValidData(Feedback date)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:21]: rename

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            var currentData = DateTime.Now.Date;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:23]: currentDat**e**

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            var feedbackData = date.CreatedOn;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:24]: feedbackCreationDate

File: Recrutify/Recrutify.Host/Startup.cs [hunk 7]
```diff
@@ -63,7 +63,6 @@ public void ConfigureServices(IServiceCollection services)
                 .AddFluentValidation();
 
             services.AddValidators();
-
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Startup.cs:66]: return this empty line

File: Recrutify/Recrutify.DataAccess/Repositories/Abstract/ICandidateRepository.cs [hunk 8]
```diff
@@ -7,5 +7,9 @@ namespace Recrutify.DataAccess.Repositories.Abstract
...
+
+        Task<ProjectResult> GetCandidateWithProjectFeedbackAsync(Guid id, Guid projectId, Guid feedbackUserId, FeedbackType feedbackType);
```
[the remark this answers is not in the document]
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Repositories/Abstract/ICandidateRepository.cs:13]: ready

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 9]
```diff
@@ -46,10 +56,28 @@ public async Task<CandidateDTO> CreateAsync(CandidateCreateDTO candidateCreateDT
             return result;
         }
 
-        public Task UpsertFeedbackAsync(Guid id, Guid projectId, CreateFeedbackDTO feedbackDto)
+        public async Task UpsertFeedbackAsync(Guid id, Guid projectId, CreateFeedbackDTO feedbackDto)
         {
-            var feedback = _mapper.Map<Feedback>(feedbackDto);
-            return _candidateRepository.UpsertFeedbackAsync(id, projectId, feedback);
+            var candidateWithProjectFeedback = await _candidateRepository.GetCandidateWithProjectFeedbackAsync(id, projectId,
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:61]: rename projectResultWithFeedback

File: Recrutify/Recrutify.DataAccess/Repositories/Abstract/ICandidateRepository.cs (hunk 1 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Repositories/Abstract/ICandidateRepository.cs:11]: ready

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 10]
```diff
@@ -46,10 +56,28 @@ public async Task<CandidateDTO> CreateAsync(CandidateCreateDTO candidateCreateDT
...
+            if (candidateWithProjectFeedback == null)
+            {
+                var newFeedback = _mapper.Map<Feedback>(feedbackDto);
+                newFeedback.CreatedOn = DateTime.UtcNow;
+                await _candidateRepository.UpsertFeedbackAsync(id, projectId, newFeedback);
+            }
+            else
+            {
+                var copy = candidateWithProjectFeedback.DeepCopy();
+                var feedbackToUpdate = _mapper.Map(feedbackDto, copy);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:72]: pass candidateWithProjectFeedback.DeepCopy() in Map method, don't declare var copy

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 11]
```diff
@@ -46,10 +56,28 @@ public async Task<CandidateDTO> CreateAsync(CandidateCreateDTO candidateCreateDT
...
+                var isvalid = await _validator.ValidateAsync(feedbackToUpdate);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:73]: isValid

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs (hunk 2 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:37]: ready

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        public ProjectResultValidator()
+        {
+            RuleFor(f => f.Feedbacks.FirstOrDefault())
+                .NotNull()
+                .Must(BeAValidData)
+                .WithMessage("Time for editing is over");
+            RuleFor(f => f.Status)
+                .IsInEnum()
+                .Must(s => !new[] { Status.Accepted, Status.Declined, Status.WaitingList, }.Contains(s))
+                .WithMessage("Status is not available for update");
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:19]: Feedback is not available for update for candidate in current status

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+namespace Recrutify.Services.Validators
+{
+    public class ProjectResultValidator : AbstractValidator<ProjectResult>
+    {
+        public ProjectResultValidator()
+        {
+            RuleFor(f => f.Feedbacks.FirstOrDefault())
+                .NotNull()
+                .Must(BeAValidData)
+                .WithMessage("Time for editing is over");
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:15]: Сannot be updated from one day after creation

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs (hunk 4 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:22]: ready

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs (hunk 5 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:24]: ready

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs (hunk 6 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:25]: done

File: Recrutify/Recrutify.Host/Startup.cs (hunk 7 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Startup.cs:66]: done

File: Recrutify/Recrutify.Services/Recrutify.Services.csproj [hunk 14]
```diff
@@ -6,6 +6,7 @@
 
   <ItemGroup>
     <PackageReference Include="AutoMapper" Version="10.1.1" />
+    <PackageReference Include="DocumentFormat.OpenXml" Version="2.14.0" />
```
[the remark this answers is not in the document]
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Recrutify.Services.csproj:9]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 9 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 10 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs:72]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 11 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs:73]: done

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs (hunk 12 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:19]: done

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs (hunk 13 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:15]: done

**Pull request merged**